include/dr/mhp/algorithms/sort.hpp

```diff
@@ -0,0 +1,304 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#ifdef SYCL_LANGUAGE_VERSION
+#include <dr/shp/init.hpp>
+#include <oneapi/dpl/algorithm>
+#include <oneapi/dpl/execution>
+#include <oneapi/dpl/iterator>
+#endif
+
+#include <mpi.h>
+
+#include <algorithm>
+#include <utility>
+
+#include <dr/concepts/concepts.hpp>
+#include <dr/detail/logger.hpp>
+#include <dr/detail/onedpl_direct_iterator.hpp>
+#include <dr/detail/ranges_shim.hpp>
+#include <dr/mhp/global.hpp>
+
+namespace dr::mhp {
+
+namespace __detail {
+
+template <rng::forward_range R, typename Compare>
+void local_sort(R &r, Compare &&comp) {
+  if (rng::size(r) >= 2) {
+    if (mhp::use_sycl()) {
+#ifdef SYCL_LANGUAGE_VERSION
+      auto policy = dpl_policy();
+      auto &&local_segment = dr::ranges::__detail::local(r);
+      drlog.debug("GPU dpl::sort(), size {}\n", rng::size(r));
+      oneapi::dpl::sort(
+          policy, dr::__detail::direct_iterator(rng::begin(local_segment)),
+          dr::__detail::direct_iterator(rng::end(local_segment)), comp);
+#else
+      assert(false);
+#endif
+    } else {
+      drlog.debug("cpu rng::sort, size {}\n", rng::size(r));
+      rng::sort(rng::begin(r), rng::end(r), comp);
+    }
+  }
+}
+
+template <dr::distributed_range R, typename Compare>
+void dist_sort(R &r, Compare &&comp) {
```

Comment:
this function is 240 lines long
Hard to read, e.g. because each local variable once defined is possibly used in all later code.
Please leave here TODO and create a task for refactoring this function into logical subfunctions once we fix more urgent issues like MPI bug and sort performance.